### PR TITLE
1.6.0 sync

### PR DESCRIPTION
**Pull request by isuruf**
Checklist
* [ ] Used a fork of the feedstock to propose changes
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

Closes #8
Closes #9
Fixes #10 
Closes #15

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,110 +1,104 @@
...
+  patches:
+    - nccl_socket.patch
+    - remove_shared_linker_flag_override.patch
+    - fix_std_stdint.patch
+    - fix_dispatch_apply_auto.patch
+    - fix_map_anonymous.patch
+    - cpp_extension.patch
 
 build:
   number: 0
```
Comment (hmaarrfk) [inline recipe/meta.yaml:21]:
```suggestion
  number: 0
  skip: true # [py!=38]
```

File: recipe/meta.yaml [hunk 2]
```diff
@@ -1,110 +1,104 @@
-{% set version = "1.1.0" %}
+{% set version = "1.6.0" %}
+{% set commit = "b31f58de6fa8bbda5353b3c77d9be4914399724d" %}
 
 package:
-  name: pytorch-cpu
+  name: pytorch-{{ pytorch_variant }}
```
Comment (hmaarrfk) [inline recipe/meta.yaml:5]: this is the old naming convention. Do you want to keep it, or do you want to start to follow the new convention which will install the correct pytorch with

```
conda install pytorch torchvision torchaudio cudatoolkit=10.2 -c pytorch
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,110 +1,104 @@
...
-    - make  # [unix]
+    - patch     # [not win]
+    - m2-patch  # [win]
+    - git       # [not win]
+    - m2-git    # [win]
   host:
+    - cmake
+    - git       # [not win]
+    - m2-git    # [win]
+    # GPU requirements
```
Comment (hmaarrfk) [inline recipe/meta.yaml:38]: Ah got it, it is still a WIP and you are leaving the GPU stuff commented out while the CPU mgiration is completed.

File: recipe/meta.yaml (hunk 1 above)
Comment (isuruf) [reply recipe/meta.yaml:21]: why?

Comment (hmaarrfk) [reply recipe/meta.yaml:21]: to itrate faster. but done.

**Comment by isuruf** [comment]
This is ready for a review. windows failure is expected. will have to build windows packages manually.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,110 +1,104 @@
...
-    # Maybe we should have a test for a simple Neural Network???
+  source_files:
+    - test
+  commands:
+    - OMP_NUM_THREADS=4 python ./test/run_test.py || true  # [not win]
+    - python ./test/run_test.py  # [win]
 
 about:
-  home: https://pytorch.org/
+  home: http://pytorch.org/
```
Comment (hmaarrfk) [inline recipe/meta.yaml:97]:
```suggestion
  home: https://pytorch.org/
```

File: recipe/meta.yaml [hunk 5]
```diff
@@ -1,110 +1,104 @@
...
+    #- cudnn {{ cudnn }}*              # [pytorch_variant == "gpu"]
+    #- magma                           # [pytorch_variant == "gpu"]
+    # other requirements
     - python
+    - numpy
     - pip
-    - pyyaml
-    - numpy 1.11
     - setuptools
+    - pyyaml
```
Comment (hmaarrfk) [inline recipe/meta.yaml:47]: Are all these necessary dependencies now? pytorch used to vendor everything.

File: recipe/meta.yaml [hunk 6]
```diff
@@ -1,110 +1,104 @@
...
   build:
     - {{ compiler('c') }}
     - {{ compiler('cxx') }}
-    - cmake
-    - ninja
-    - make  # [unix]
+    - patch     # [not win]
+    - m2-patch  # [win]
+    - git       # [not win]
+    - m2-git    # [win]
```
Comment (hmaarrfk) [inline recipe/meta.yaml:33]: whats the difference between `git` and `m2-git` on windows.

File: recipe/meta.yaml [hunk 7]
```diff
@@ -1,110 +1,104 @@
...
-  # Support for Windows and Python 2.7 is very experimental
-  # https://ci.appveyor.com/project/conda-forge/staged-recipes/builds/21220961#L972
-  skip: true  # [win and py27]
-  script:
-    # Most of this script is adapted from the pytorch/pythorch-cpu
-    # meta.yaml.template build.sh and bld.bat scripts
-    - {{ export }} TN_BINARY_BUILD=1
-    - {{ export }} PYTORCH_BINARY_BUILD=1
-    - {{ export }} NO_CUDA=1
-    - {{ export }} PYTORCH_BUILD_VERSION={{ PKG_VERSION }}
```
Comment (hmaarrfk) [inline recipe/meta.yaml:32]: thanks for cleanign this up.

File: recipe/meta.yaml (hunk 5 above)
Comment (isuruf) [reply recipe/meta.yaml:47]: Yes, pytorch/builder repo has this too

File: recipe/meta.yaml (hunk 6 above)
Comment (isuruf) [reply recipe/meta.yaml:33]: `git` package is broken on windows. `m2-git` is not

File: recipe/meta.yaml [hunk 8]
```diff
@@ -1,110 +1,104 @@
...
     - {{ compiler('cxx') }}
-    - cmake
-    - ninja
-    - make  # [unix]
+    - patch     # [not win]
+    - m2-patch  # [win]
+    - git       # [not win]
+    - m2-git    # [win]
   host:
+    - cmake
```
Comment (hmaarrfk) [inline recipe/meta.yaml:35]: shouldn't cmake be a build dependency?

Also for git and m2-git?

Do we want to depend on ninja?

File: recipe/meta.yaml (hunk 5 above)
Comment (hmaarrfk) [reply recipe/meta.yaml:47]: nice. yeah i noticed they finally unvendored too!

File: recipe/meta.yaml (hunk 8 above)
Comment (isuruf) [reply recipe/meta.yaml:35]: yes, but there was something weird going on in the osx build without `git` in host (as in your PR). I don't have the time to debug it. You are welcome to push changes to this PR

File: recipe/meta.yaml [hunk 9]
```diff
@@ -1,110 +1,104 @@
...
+    #- {{ pin_compatible('cudatoolkit', max_pin='x.x') }}  # [pytorch_variant == "gpu"]
+    #- {{ pin_compatible('cudnn') }}                       # [pytorch_variant == "gpu"]
+    # other requirements
     - python
     - {{ pin_compatible('numpy') }}
-    # ninja and cffi are included in
-    - ninja
     - cffi
-    - future  # [py==27]
+    - ninja
```
Comment (hmaarrfk) [inline recipe/meta.yaml:74]: can we try removing this? did you already try removing it?

File: recipe/meta.yaml (hunk 6 above)
Comment (hmaarrfk) [reply recipe/meta.yaml:33]: haha, ok.

File: recipe/meta.yaml [hunk 10]
```diff
@@ -1,110 +1,104 @@
...
-{% set export = "export" %}  # [not win]
-{% set export = "set" %}     # [win]
+  git_tag: {{ commit }}
+  patches:
+    - nccl_socket.patch
+    - remove_shared_linker_flag_override.patch
+    - fix_std_stdint.patch
+    - fix_dispatch_apply_auto.patch
+    - fix_map_anonymous.patch
+    - cpp_extension.patch
```
Comment (hmaarrfk) [inline recipe/meta.yaml:18]: Have these or similar patches made it upstream? If not, should we make PRs?

Comment (isuruf) [reply recipe/meta.yaml:18]: I don't think so. These patches are from AnacondaRecipes. It'd be good to upstream at least some of the patches.

**Comment by hmaarrfk** [comment]
This looks really good, can we:

- [ ] Use https for the url
- [ ] Add a comment as to why we are using `m2-git` on windows (i.e. because it isn't broken as of Dec 2020)
- [ ] Add a comment as to why we are including cmake in the host section
- [ ] Add a comment as to why we are depending on ninja as a runtime requirement


I've checked that for now, we still create a package called pytorch-cpu that is distinct from the upstream more common package name of `pytorch` meaning that we are still "distinct" from the upstream instructions.

**Comment by isuruf** [comment]
I'm sorry I don't have the time. Maybe someone else can take over the PR now that we have the build passing on CI

**Comment by hmaarrfk** [comment]
Sure, most of this is cleanup. happy to give it a stab eventually, just if we merge, we will never find the time.

File: recipe/build.sh [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+export USE_NINJA=OFF
+export INSTALL_TEST=0
+
+# MacOS build is simple, and will not be for CUDA
+if [[ "$OSTYPE" == "darwin"* ]]; then
+    python -m pip install . --no-deps -vv
+    exit 0
+fi
+
+# std=c++11 is required to compile some .cu files
```
Comment (hmaarrfk) [inline recipe/build.sh:52]: c++11 or c++14?

Comment (isuruf) [reply recipe/build.sh:52]: 14

**Comment by hmaarrfk** [comment]
- [ ] Add back ninja as a run dependency if linux and macosx builds fail.

**Comment by hmaarrfk** [comment]
Thanks @isuruf this is amazing!

**Comment by hmaarrfk** [comment]
Thanks. @sodre did you have any other comments? otherwise i'm going to try to merge by the end of the day.

File: recipe/meta.yaml (hunk 6 above)
Comment (dhirschfeld) [reply recipe/meta.yaml:40]: @isuruf - any pointers or could you elaborate what in particular is broken about `git` on Windows? If there's an issue I can test and comment.

**Comment by hmaarrfk** [comment]
Thanks @isuruf

**Pull request merged**